File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        res = ""
        resLen = 0

        # helper function: expand around center
        def expandAroundCenter(l: int, r: int):
            nonlocal res, resLen
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > resLen:
                    res = s[l:r+1]
                    resLen = (r - l + 1)
                l -= 1
                r += 1

        for i in range(len(s)):
            # Odd length palindrome
            expandAroundCenter(i, i)
            # Even length palindrome
            expandAroundCenter(i, i+1)

        return res

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    if (s.length < 2) return s;

    let start = 0;
    let maxLength = 1;

    function expand(left, right) {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            let currentLength = right - left + 1;
            if (currentLength > maxLength) {
                start = left;
                maxLength = currentLength;
            }
            left--;
            right++;
        }
    }

    for (let i = 0; i < s.length; i++) {
        // Odd length palindrome (center at i)
        expand(i, i);

        // Even length palindrome (center between i and i+1)
        expand(i, i + 1);
    }

    return s.substring(start, start + maxLength);
};